File: Numerical_Methods/Solving Ordinary Differential Equations (ODEs)/Euler’s Method/java_solutions/Problem2.java
public class Problem2 {
    public static void main(String[] args) {
        double h = 0.2;  // Step size
        double x0 = 0.0; // Initial x
        double y0 = 1.0; // Initial y
        double targetX = 1.0; // Target x value
        
        System.out.println("Euler's Method Solution for Problem 2");
        System.out.println("y' = 2x - y, y(0) = 1");
        System.out.println("Step size (h) = " + h);
        System.out.println("\nx\t\ty\t\tExact\t\tError");
        System.out.println("----------------------------------------");
        
        double x = x0;
        double y = y0;
        
        while (x <= targetX) {
            // Calculate exact solution: y = 2x - 2 + 3e^(-x)
            double exact = 2 * x - 2 + 3 * Math.exp(-x);
            double error = Math.abs(exact - y);
            
            System.out.printf("%.4f\t\t%.4f\t\t%.4f\t\t%.4f%n", x, y, exact, error);
            
            if (x == targetX) break;
            
            // Calculate next y using Euler's method
            double slope = 2 * x - y;  // f(x,y) = 2x - y
            y = y + h * slope;
            x = x + h;
        }
    }
} 

File: Numerical_Methods/Solving Ordinary Differential Equations (ODEs)/Euler’s Method/python_solutions/problem2.py
import numpy as np

def euler_method(h, x0, y0, target_x):
    x = x0
    y = y0
    
    print("Euler's Method Solution for Problem 2")
    print("y' = 2x - y, y(0) = 1")
    print(f"Step size (h) = {h}")
    print("\nx\t\ty\t\tExact\t\tError")
    print("----------------------------------------")
    
    while x <= target_x:
        # Calculate exact solution: y = 2x - 2 + 3e^(-x)
        exact = 2 * x - 2 + 3 * np.exp(-x)
        error = abs(exact - y)
        
        print(f"{x:.4f}\t\t{y:.4f}\t\t{exact:.4f}\t\t{error:.4f}")
        
        if x == target_x:
            break
            
        # Calculate next y using Euler's method
        slope = 2 * x - y  # f(x,y) = 2x - y
        y = y + h * slope
        x = x + h

if __name__ == "__main__":
    h = 0.2      # Step size
    x0 = 0.0     # Initial x
    y0 = 1.0     # Initial y
    target_x = 1.0  # Target x value
    
    euler_method(h, x0, y0, target_x) 